engine/classes/Elgg/Cache/DiDefinitionCache.php

```diff
@@ -0,0 +1,156 @@
+ * notice, this list of conditions and the following disclaimer.
+ *  * Redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution.
+ *  * Neither the name of the Stash Project nor the
+ * names of its contributors may be used to endorse or promote products
+ * derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL Robert Hafner BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @access private
+ * @since 3.0.0
+ */
+class DiDefinitionCache implements DoctrineCache {
+
+	/**
+	 * @var Pool
+	 */
+	protected $pool;
+
+	/**
+	 * Constructor
+	 *
+	 * @param Pool $pool Stash pool
+	 */
+	public function __construct(Pool $pool) {
+		$this->pool = $pool;
+		$this->pool->setNamespace('di_definitions');
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function fetch($id) {
+		$item = $this->pool->getItem($id);
+		$value = $item->get();
+		return $item->isMiss() ? false : $value;
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function contains($id) {
+		$item = $this->pool->getItem($id);
+		return !$item->isMiss();
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function save($id, $data, $lifeTime = 0) {
+		if ($lifeTime === 0) {
+			$lifeTime = null;
+		}
+		$item = $this->pool->getItem($id);
+		$item->set($data);
+		$item->setTTL($lifeTime);
+		return $this->pool->save($item);
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function delete($id) {
+		return $this->pool->deleteItem($id);
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function getStats() {
+		$stats = array();
+		$stats['memory_usage'] = 'NA';
+		$stats['memory_available'] = 'NA';
+		$stats['uptime'] = 'NA';
+		$stats['hits'] = 'NA';
+		$stats['misses'] = 'NA';
+		return $stats;
+	}
+
+	/**
+	 * Clear the cache
+	 */
+	public function flush() {
+		$this->pool->clear();
+	}
+
+	/**
+	 * Build a cache
+	 *
+	 * @param Config $config Elgg config
+	 * @return DiDefinitionCache|null
+	 */
+	public static function factory(Config $config) {
+		$driver = call_user_func(function () use ($config) {
+			if (is_callable('apcu_fetch')) {
+				return new Apc();
+			}
+
+			$has_class = class_exists('Memcache') || class_exists('Memcached');
```

Comment:
Hmm, I think this should just be APC like PHP-DI recommends. Filenames stored in cache might not match between servers and even with the namespace we can't flush just the DI entries, so we don't want to have to flush the whole redis/memcache pools, right?